hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py

```diff
@@ -119,25 +133,10 @@ async def __aexit__(
     ) -> None:
         assert self._task is not None
 
-        if not self._task.done():
-            if exc_val:
-                self._task.cancel()
-                try:
-                    value = await self._task
-                    if self._closable:
-                        value.close()
-                except:
-                    _, exc, _ = sys.exc_info()
-                    if exc is not exc_val:
-                        log.warning('dropping preempted task exception', exc_info=True)
-            else:
-                value = await self._task
-                if self._closable:
-                    value.close()
+        if self._closable and self._task.done() and not self._task.cancelled():
+            (await self._task).close()
```

Comment:
Do we not need to close the task result in the case the task is not done and we cleanup the future below in the else clause?

Reply:
If the task is not done, then we will cancel it in `_cleanup_future` before it gets a chance to return anything (closable or not).

The only way a task could return a closable would be to catch CancelledError, ignore it, and then return something anyway. We *must not* write tasks like this. That's why we use this pattern everywhere:
```python3
except CancelledError:
    raise
except Exception as exc:
    # ...
```

It is our job to ensure our tasks *never* ignore CancelledError. If you receive a CancelledError, you *must* terminate (by raising CancelledError) as soon as possible. The task can (quickly) cleanup resources but it definitely shouldn't `return` and it really shouldn't do any compute intensive work.